towerlib/towerlib.py

```diff
@@ -2018,3 +2031,178 @@ def get_schedule_by_name(self, name):
+
+        Args:
+            scm_url: the URL of the relevant repository configured in the project.
+            branch_name: the name of the branch, which is selected as scm_branch parameter of the project.
+            organization_name: the name of the organization.
+
+        """
+        organization = self.get_organization_by_name(organization_name)
+        if not organization:
+            raise InvalidOrganization(organization_name)
+        matching_projects = (project for project in organization.projects if all([project.scm_url == scm_url,
+                                                                                  project.scm_branch == branch_name]))
+        outputs = []
+        for project in matching_projects:
+            output = project.update()
+            if not output:
+                self.logger.error(f'{project.name} failed to update')
+            outputs.append(output)
+        return all(outputs)
+
+    def get_jobs_by_name(self, name):
+        """Get filtered list of jobs for a given name.
+
+        Args:
+            name: the given job name.
+
+        Returns:
+             list: the filtered list of jobs.
+
+        """
+
+        return self.jobs.filter({'name__iexact': name})
+
+    def get_job_by_id(self, id_):
+        """Retrieves a job by id.
+
+        Args:
+            id_: The id of the job to retrieve.
+
+        Returns:
+            Host: The host if a match is found else None.
+
+        """
+        return next(self.jobs.filter({'id': id_}), None)
+
+    def get_project_update_by_id(self, id_):
+        """Retrieves a project_update by id.
+
+        Args:
+            id_: The id of the project_update to retrieve.
+
+        Returns:
+            Host: The project_update if a match is found else None.
+
+        """
+        return next(self.project_updates.filter({'id': id_}), None)
+
+    def get_project_updates_by_name(self, name):
+        """Retrieves project_updates matching a certain name.
+
+        Args:
+            name: the given job_update name.
+
+        Returns:
+             list: the filtered list of project update jobs.
+
+        """
+        return self.project_updates.filter({'name__iexact': name})
+
+    @property
+    def job_events(self):
+        """The job templates configured in tower.
+
+        Returns:
+            EntityManager: The manager object for job templates.
+
+        """
+        return EntityManager(self,
+                             entity_name='job_events',
+                             entity_object='JobEvent',
+                             primary_match_field='name')
+
+    def get_all_groups_by_host_id(self, host_id):
+        """Get groups for a particular host, which are directly and indirectly connected.
+
+        Args:
+            host_id: the id of the given host..
+
+        Returns:
+            list: list of custom groups.
+
+        """
+        url = "{api}/hosts/{id}/all_groups/".format(api=self.api, id=host_id)
```

Comment:
Since you have implemented the `all_groups` attributes on the host it would be much better to return those which are modeled objects instead of a raw json like here. 
code could be:
```
host = self.get_host_by_id(host_id)
if not host:
    return []
return host.all_groups
```